presto-main/src/main/java/com/facebook/presto/dispatcher/LocalDispatchQuery.java

```diff
@@ -71,14 +73,16 @@ public LocalDispatchQuery(
             Consumer<QueryExecution> querySubmitter)
     {
         this.stateMachine = requireNonNull(stateMachine, "stateMachine is null");
+        this.queryMonitor = requireNonNull(queryMonitor, "queryMonitor is null");
         this.queryExecutionFuture = requireNonNull(queryExecutionFuture, "queryExecutionFuture is null");
         this.clusterSizeMonitor = requireNonNull(clusterSizeMonitor, "clusterSizeMonitor is null");
         this.queryExecutor = requireNonNull(queryExecutor, "queryExecutor is null");
         this.querySubmitter = requireNonNull(querySubmitter, "querySubmitter is null");
 
         addExceptionCallback(queryExecutionFuture, throwable -> {
-            stateMachine.transitionToFailed(throwable);
-            queryMonitor.queryImmediateFailureEvent(stateMachine.getBasicQueryInfo(Optional.empty()), toFailure(throwable));
+            if (stateMachine.transitionToFailed(throwable)) {
+                queryMonitor.queryImmediateFailureEvent(stateMachine.getBasicQueryInfo(Optional.empty()), toFailure(throwable));
+            }
```

Comment:
can this just invoke the `fail` function?

Reply:
I was wary of referencing a method in the constructor before the object is fully initialized.